### PR TITLE
Wfh tomorrow

### PR DESCRIPTION
**Pull request by TNagaraja**
Added functionality so users can optionally choose to specify wfh date and/or time interval. '/wfh MM-DD-YYYY time1 to time2'. Also accepts 'tomorrow' to specify tomorrow's date.

### CODE REVIEW

File: test/parse-text.spec.js
Comment (MarkLeMerise) [inline test/parse-text.spec.js]: Are you actually executing `act()` twice in this suite? Once here, and then again on line 143?

File: src/google-calendar-api.js
Comment (MarkLeMerise) [inline src/google-calendar-api.js]: This is a good update and actually makes testing easier, but I'm thinking we should move this Google auth block to its own module. Thoughts?

File: src/google-calendar-api.js
Comment (MarkLeMerise) [inline src/google-calendar-api.js]: Well, I just learned something: I didn't know you could reference other parameters in default parameters. That's really cool! In transpiled code, this works implicitly, but I was surprised here so much so that I tried it in the console.

Though it's "compact", I would reconsider the readability of this function signature, a lot is happening in a single line. Maybe moving some of this defaulting into the function body itself would improve readability.

Alternatively, you can define the defaulting in a separate function so your signature looks like this:
```
const plusOneDay = date => new Date(moment(date).add(1, 'day').startOf('day').format());
checkIfWfhEventExists(employeeName, eventStartDateTime, eventEndDateTime = plusOneDay(eventStartDateTime)) { /* ... */ }
```

Though even that looks a bit verbose. Ultimately, it's a matter of opinion as these are all equivalent, but readability is important to other developers who aren't as familiar with the code.

File: test/google-calendar-api.spec.js
Comment (MarkLeMerise) [inline test/google-calendar-api.spec.js]: I think [`chance.date` returns a `Date` object](http://chancejs.com/#date). You probably don't need to wrap it.

File: test/google-calendar-api.spec.js
Comment (MarkLeMerise) [inline test/google-calendar-api.spec.js]: Sinon's API is specialized for asserting conditions about fake methods (i.e. "spies" & "mocks", in this case, `fakeCalendarApi.list` is a mock), so here you can use that API like `sinon.assert.calledWith(fakeCalendarApi.list, sinon.match({ timeMax': moment(endDateTime).format()))`. I would recommend reading the documentation to get a full gist of what Sinon has to offer.

Comment (MarkLeMerise) [reply test/google-calendar-api.spec.js]: Also, since you have two `moment` objects, I think you can compare them outright instead of having to `format()` then and compare the strings.

File: src/parse-text.js
Comment (MarkLeMerise) [inline src/parse-text.js]: Couple things about this method:
1. Case-insenstitive string comparisons are less prone to break. For example, consider if I type `/wfh TOMORROW`, would this logic still work?
2. To make the method more compact, you can return the comparison directly as the result like `return text.toLowerCase().includes('tomorrow')`.

File: src/index.js
Comment (MarkLeMerise) [inline src/index.js:11]: From a user experience perspective, I'm a little confused about the purpose of "clear". Is the intention that I type `/wfh clear` and that deletes my WFH event for the current day?

File: test/google-calendar-api.spec.js
Comment (MarkLeMerise) [inline test/google-calendar-api.spec.js]: Sinon has some nice sugar for this type of stub: `sinon.stub().resolves({})` 😉

**Review by MarkLeMerise — CHANGES_REQUESTED** [review]

**Comment by MarkLeMerise** [comment]
Here are a few higher-level thoughts I thought of that you might want to consider as you re-read through your code:
1. Many places in the logic, there is a lot of "weaving" in and out of `moment` objects, and the native `Date` format. Consistency is important here, especially given that date parsing is a pillar of this feature. If I were to make a recommendation, I would never even touch the native `Date` object, as the `moment` library is a battle-tested wrapper, which shields the developer from common `Date` pitfalls.
2. You've got a good handle on parsing the date, but remember, this text comes _directly_ from a user, and users screw up...**a lot**, so it's important to consider "edge" cases in testing. Consider this user experience, what happens if someone types one too many spaces between the times? (Or forgets spaces). Knowing when to "forgive" the input and do your best to parse it and knowing when to throw up in your hands and say this can't be parsed is an important balance to find. In general, I find user input parsing and searching to be one of the more difficult programming patterns to get right. 
3. Code readability is an important aspect of development that should not be overlooked. As a reviewer or fellow developer, I may or may not run your code. More than likely, I'm going to scan through to "execute in my head". Think of code authoring like authoring a book. A paragraph-long sentence may seem eloquent, but by the time I'm done, I probably won't grok the full thought. For example, instead of explicitly returning `true` or `false`, ask yourself can I simply return the expression which produces that `boolean` result? Developing this sense is a lifetime skill and you won't always please everyone, but we do our best to use tools (like ESLint) to keep the look & feel consistent.

File: test/parse-text.spec.js
Comment (TNagaraja) [reply test/parse-text.spec.js]: Yeah, that was unnecessary - removing line 140.

File: src/google-calendar-api.js
Comment (TNagaraja) [reply src/google-calendar-api.js]: I think that might be a good idea if we wanted to reference the Google API in new classes. What's your reasoning for moving the block to its own module?

File: src/google-calendar-api.js
Comment (TNagaraja) [reply src/google-calendar-api.js]: I agree that this change would improve readability and is worth implementing

File: test/google-calendar-api.spec.js
Comment (TNagaraja) [reply test/google-calendar-api.spec.js]: Cool stuff. I'll add that in.

File: test/google-calendar-api.spec.js
Comment (TNagaraja) [reply test/google-calendar-api.spec.js]: Seems to be the case. I think your comment for using moment across the board instead of native Date objects is a good idea. I don't think moment has a random generator so I'll probably construct moment objects in the tester classes by supplying chance.date

File: test/google-calendar-api.spec.js
Comment (TNagaraja) [reply test/google-calendar-api.spec.js]: I agree it's cleaner just to compare the two moment objects outright

File: src/parse-text.js
Comment (TNagaraja) [reply src/parse-text.js]: This is much better. I'll change it.

File: src/index.js
Comment (TNagaraja) [reply src/index.js:11]: That's correct. It deletes the WFH event for the specified date. Typing "/wfh clear" will clear the current date's wfh event. It's also valid to type "/wfh tomorrow clear" or "/wfh MM-DD-YYYY clear."

Comment (MarkLeMerise) [reply src/index.js:11]: Okay, thanks for the clarification. Now that we have this "clearing" functionality, we should probably remove any code paths that "toggled" the WFH event since `/wfh clear` is the equivalent of typing `/wfh` a second time.

File: src/google-calendar-api.js
Comment (MarkLeMerise) [reply src/google-calendar-api.js]: This module is really concerned with dealing with the Google Calendar API only. This block deals with authorizing for _any_ Google API. To me, it would improve readability and testing not only of that module but this one as well, as the `credentials` object could more easily be mocked.

File: src/index.js
Comment (TNagaraja) [reply src/index.js:11]: I was thinking the toggle should still be the primary way to set/unset a wfh event. I wanted to keep `/wfh clear` in the code in the event that somebody is uncertain if they have a wfh event on a given day and want to be certain that they can clear the wfh for that day. If `/wfh clear` causes confusion, it might be better to remove that feature than to remove the toggle.

File: test/google-calendar-api.spec.js
Comment (TNagaraja) [reply test/google-calendar-api.spec.js]: I tried pulling out the `.format` and `sinon.assert.match` does not know how to compare the two moment objects. I think I'll have to compare the strings produced by `.format`

Comment (MarkLeMerise) [reply test/google-calendar-api.spec.js]: Sinon wouldn't know how to compare them, but [`moment` itself has built-in methods for comparing two `moment` objects](https://momentjs.com/docs/#/query/is-same/), so you can do something like `sinon.assert.match(moment.isSame(m1, m2), true)`.

File: src/index.js
Comment (MarkLeMerise) [reply src/index.js:11]: The WFH calendar is (in the eyes of Symphono employees) only writeable via this command's interface, so I think you were right to include clearing functionality because if I set a far future WFH, I can literally only clear it through this command.  However, it might be a confusing user experience (UX) to provide two different ways to clear a WFH event. Consider this flow:
1. User types `/wfh` -> server says "WFH created for today".
2. User types `/wfh` -> server says "WFH deleted for today."
3. User types `/wfh tomorrow` -> server says "WFH created for tomorrow".
4. User types `/wfh tomorrow` -> server says "WFH created for tomorrow."

See how the experiences are "asymmetrical"? Rather, it might be more helpful to tell the user they're already on the calendar for that day:

1. User types `/wfh` -> server says "WFH created for today"
2. User types `/wfh` -> server says "You're already WFH today"
3. User types `/wfh clear` -> server says "WFH deleted for today"
4. User types `/wfh tomorrow` -> server says "WFH created for tomorrow"
5. User types `/wfh tomorrow` -> server says "You're already WFH tomorrow"
6. User types `/wfh tomorrow clear` -> server says "WFH deleted for tomorrow"

Comment (TNagaraja) [reply src/index.js:11]: Currently, if a user types `/wfh tomorrow` twice - the server will delete the event for that day and send a similar message back to slack saying they are going to the office on tomorrow's date after the second command. The toggle works correctly regardless of date or time interval so I don't think the experiences are asymmetrical. You do not have to use `/wfh clear tomorrow` to remove tomorrow's wfh event. However, if you think that it is more intuitive to replace the toggle function with a function that adds a wfh event if one is not there but does not handle deletion, it may make sense to make this change.

File: src/google-calendar-api.js
Comment (TNagaraja) [reply src/google-calendar-api.js]: Makes sense. I'll include that change.

File: test/google-calendar-api.spec.js
Comment (TNagaraja) [reply test/google-calendar-api.spec.js]: Change implemented in next push

File: src/index.js
Comment (MarkLeMerise) [reply src/index.js:11]: Okay, my understanding was a bit flawed then, thanks for clarifying. I'm okay with what you have then. We'll just need to update the command usage documentation in Slack once this merges.

Comment (MiguelCastillo) [reply src/index.js:11]: hmmm - i personally think that calling `/wfh tomorrow` twice should not delete the event...  I personally like explicit actions that are destructive.

Comment (TNagaraja) [reply src/index.js:11]: I think making deleting a wfh event an explicit action is better practice. However, do you think that such a change would not be user friendly because people are accustomed to using `\wfh` as a toggle?

Comment (MiguelCastillo) [reply src/index.js:11]: i think its fine to change the functionality as long as we communicate that...  but maybe @MarkLeMerise can shed some light as to why it was done that way...?  And it if would be weird to change it...?

Comment (MarkLeMerise) [reply src/index.js:11]: Historically, this was written to be a simple command which took no arguments, thus no explicit "clear" command. Now that we take arguments, I sorta see the merits of both arguments. @TNagaraja I like that you're thinking about how changing the toggle "breaks the public API" of this command, but @MiguelCastillo also makes a good point about making the user aware of a "destructive" action. In the spirit of code reviews, @TNagaraja, I'm going to let you make the final decision. Thankfully, since this is an internal tool, we only have a small user base that will need to be informed about any changes to this command.

Comment (TNagaraja) [reply src/index.js:11]: I think it makes sense to make the delete explicit in that case. It would make the user more clearly aware of their actions.

Comment (TNagaraja) [reply src/index.js:11]: Changes in latest commit

**Review by MarkLeMerise — APPROVED** [review]

**Pull request merged**
